### PR TITLE
perf(lexer): jump table

### PR DESCRIPTION
**Pull request by Boshen**
To review this PR, scroll to the bottom of the file changes.

---

So I always wanted to apply the jump table method from ratel https://github.com/ratel-rust/ratel-core/blob/master/ratel/src/lexer/mod.rs

CPU instructions count should be greatly reduced because we only do keyword match on the ascii characters a-z (previously it was checking for keyword for all identifiers). There is no hot spot for keyword match anymore. 

The benchmark numbers are not significant, probably need to find a better way to indicate this actually improves things.

### CODE REVIEW

**Comment by Boshen** [comment]
I see two further optimizations worth trying out, but I'll leave this to another time.

1. Tighter loop for skipping white spaces?

2.
```rust
            if let Some(c) = self.current.chars.clone().next() {
                let kind = self.match_char(c);
```

This peek may be unnecessary, can be removed after some refactoring.

**Comment by Boshen** [comment]
The benchmark numbers are too subtle, but this PR should reduce the total number of CPU instructions for the matching keyword case.

**Comment by Boshen** [comment]
Profiler shows cost of building BTreeMap for trivias is significant on comment heavy files such as typescript.js.

File: crates/oxc_parser/src/lexer/mod.rs [hunk 1]
```diff
@@ -1478,3 +1289,526 @@ enum SurrogatePair {
...
+    "await" => Kind::Await,
+    _ => Kind::Ident,
+};
+
+const L_B: ByteHandler = |lexer| match lexer.identifier_name_handler() {
+    "bigint" => Kind::BigInt,
+    "boolean" => Kind::Boolean,
+    "break" => Kind::Break,
+    _ => Kind::Ident,
+};
```
Comment (strager) [inline crates/oxc_parser/src/lexer/mod.rs:1665]: Can't you skip the first byte?
```suggestion
const L_B: ByteHandler = |lexer| match lexer.identifier_name_handler()[1:] {
    "igint" => Kind::BigInt,
    "oolean" => Kind::Boolean,
    "reak" => Kind::Break,
    _ => Kind::Ident,
};
```

**Comment by strager** [comment]
> There is no hot spot for keyword match anymore.

I think that's because you spread the heat across 21 different functions! If you add up the cost of `L_A`, `L_B`, etc. I think it'd be the same as keyword recognition before this patch.

**Comment by strager** [comment]
> this PR should reduce the total number of CPU instructions

I observe this to be true on my M1 (AArch64):
```
# targe0: Git commit b2db6f7ef14d9beee842107f8502f227b31ebfc7
# target: Git commit c8a215ea5ece9f73f3013af07a9c7fa3ffbc4557

$ for target in targe0 target ; do echo $target ; for _ in {1..10} ; do command time -l $target/release/oxlint --threads=2 -Aall ~/tmp/jquery-3.5.1.js |& grep 'instructions retired' ; done ; done
targe0
            74039247  instructions retired
            74154892  instructions retired
            74241578  instructions retired
            74076408  instructions retired
            74078166  instructions retired
            73936371  instructions retired
            74105009  instructions retired
            73900220  instructions retired
            73988127  instructions retired
            73985852  instructions retired
target
            72415843  instructions retired
            72417568  instructions retired
            72666247  instructions retired
            72534572  instructions retired
            72442302  instructions retired
            72377676  instructions retired
            72470803  instructions retired
            72491875  instructions retired
            72404827  instructions retired
            72416905  instructions retired
```

**Comment by Boshen** [comment]
> 

Previously it was running on all identifiers - redundant for things like uppercase letters. It's only 21 now, so much better!

> I observe this to be true on my M1 (AArch64):

TIL how to count cpu instructions, thank you!

**Pull request merged**